### ISSUE TITLE
Can't find input_layer after freezing graph for Cifar10 model (it uses shuffle queue)

### ISSUE THREAD

**Issue by drag0** [tensorflow/tensorflow#5464]
I've run [cifar10_train.py]( https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/image/cifar10/cifar10_train.py
) example for some time and saved graph and checkpoint files. After that I used freeze_graph.py from python/tools and it worked fine, no errors. Later I've built [label_image](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/label_image) example, it worked fine with it's predefined values, but when I tried to use frozen graph and provided different flags, it failed because I couldn't find proper --input_layer value. I've tried many different values but nothing worked.

I've tried most of operations as --input_layer in this code snippet from Cifar10 example but none worked for me:

```
def read_cifar10(filename_queue):
  """Reads and parses examples from CIFAR10 data files.

  Recommendation: if you want N-way read parallelism, call this function
  N times.  This will give you N independent Readers reading different
  files & positions within those files, which will give better mixing of
  examples.

  Args:
    filename_queue: A queue of strings with the filenames to read from.

  Returns:
    An object representing a single example, with the following fields:
      height: number of rows in the result (32)
      width: number of columns in the result (32)
      depth: number of color channels in the result (3)
      key: a scalar string Tensor describing the filename & record number
        for this example.
      label: an int32 Tensor with the label in the range 0..9.
      uint8image: a [height, width, depth] uint8 Tensor with the image data
  """

  class CIFAR10Record(object):
    pass
  result = CIFAR10Record()

  # Dimensions of the images in the CIFAR-10 dataset.
  # See http://www.cs.toronto.edu/~kriz/cifar.html for a description of the
  # input format.
  label_bytes = 1  # 2 for CIFAR-100
  result.height = 32
  result.width = 32
  result.depth = 3
  image_bytes = result.height * result.width * result.depth
  # Every record consists of a label followed by the image, with a
  # fixed number of bytes for each.
  record_bytes = label_bytes + image_bytes

  # Read a record, getting filenames from the filename_queue.  No
  # header or footer in the CIFAR-10 format, so we leave header_bytes
  # and footer_bytes at their default of 0.
  reader = tf.FixedLengthRecordReader(record_bytes=record_bytes)
  result.key, value = reader.read(filename_queue)

  # Convert from a string to a vector of uint8 that is record_bytes long.
  record_bytes = tf.decode_raw(value, tf.uint8, name="input_layer_node")

  # The first bytes represent the label, which we convert from uint8->int32.
  result.label = tf.cast(
      tf.slice(record_bytes, [0], [label_bytes]), tf.int32)

  # The remaining bytes after the label represent the image, which we reshape
  # from [depth * height * width] to [depth, height, width].
  depth_major = tf.reshape(tf.slice(record_bytes, [label_bytes], [image_bytes]),
                           [result.depth, result.height, result.width])
  # Convert from [depth, height, width] to [height, width, depth].
  result.uint8image = tf.transpose(depth_major, [1, 2, 0])

  return result
```

Anyone had success with this one?

>### Environment info
>Operating System: Ubuntu 14.04
>CUDA: 8.0
>cuDNN: 5.0
>tf.__version__
'0.11.0rc2'

**Comment by drag0** [comment]
By using first convolutional layer as input (conv1/Conv2D) I get an interesting error:

> E tensorflow/examples/label_image/main.cc:306] Running model failed: Invalid argument: Must provide as many biases as the last dimension of the input tensor: [64] vs. [1,32,32,3]
>          [[Node: conv1/BiasAdd = BiasAdd[T=DT_FLOAT, data_format="NHWC", _device="/job:localhost/replica:0/task:0/cpu:0"](_recv_conv1/Conv2D_0, conv1/biases/read)]]

It is a bit strange to me, I've checked label_image example and it uses convention for image ops [batch_size, img_width, img_height, channels] and Cifar10 example also uses the same. Now, if Cifar10 is training and running properly, why do we get this error here? It doesn't make sense to me. 

Here is code from label_image example that reads image as a tensor:

```
// Given an image file name, read in the data, try to decode it as an image,
// resize it to the requested size, and then scale the values as desired.
Status ReadTensorFromImageFile(string file_name, const int input_height,
                               const int input_width, const float input_mean,
                               const float input_std,
                               std::vector<Tensor>* out_tensors) {
  auto root = tensorflow::Scope::NewRootScope();
  using namespace ::tensorflow::ops;  // NOLINT(build/namespaces)

  string input_name = "file_reader";
  string output_name = "normalized";
  auto file_reader = ReadFile(root.WithOpName(input_name), file_name);
  // Now try to figure out what kind of file it is and decode it.
  const int wanted_channels = 3;
  Output image_reader;
  if (tensorflow::StringPiece(file_name).ends_with(".png")) {
    image_reader = DecodePng(root.WithOpName("png_reader"), file_reader,
                             DecodePng::Channels(wanted_channels));
  } else if (tensorflow::StringPiece(file_name).ends_with(".gif")) {
    image_reader = DecodeGif(root.WithOpName("gif_reader"), file_reader);
  } else {
    // Assume if it's neither a PNG nor a GIF then it must be a JPEG.
    image_reader = DecodeJpeg(root.WithOpName("jpeg_reader"), file_reader,
                              DecodeJpeg::Channels(wanted_channels));
  }
  // Now cast the image data to float so we can do normal math on it.
  auto float_caster =
      Cast(root.WithOpName("float_caster"), image_reader, tensorflow::DT_FLOAT);
  // The convention for image ops in TensorFlow is that all images are expected
  // to be in batches, so that they're four-dimensional arrays with indices of
  // [batch, height, width, channel]. Because we only have a single image, we
  // have to add a batch dimension of 1 to the start with ExpandDims().
  auto dims_expander = ExpandDims(root, float_caster, 0);
  // Bilinearly resize the image to fit the required dimensions.
  auto resized = ResizeBilinear(
      root, dims_expander,
      Const(root.WithOpName("size"), {input_height, input_width}));
  // Subtract the mean and divide by the scale.
  Div(root.WithOpName(output_name), Sub(root, resized, {input_mean}),
      {input_std});

  // This runs the GraphDef network definition that we've just constructed, and
  // returns the results in the output tensor.
  tensorflow::GraphDef graph;
  TF_RETURN_IF_ERROR(root.ToGraphDef(&graph));

  std::unique_ptr<tensorflow::Session> session(
      tensorflow::NewSession(tensorflow::SessionOptions()));
  TF_RETURN_IF_ERROR(session->Create(graph));
  TF_RETURN_IF_ERROR(session->Run({}, {output_name}, {}, out_tensors));
  return Status::OK();
}
```

**Comment by sherrym** [comment]
Could you please include the exact steps, and the exact code if you have written any?

Thanks,
Sherry

**Comment by drag0** [comment]
@sherrym - ok, the thing was that queue returns images in batches of 128 images at a time so CNN expects tensor of shape [batch_size, height, width, channels]. When I changed batch_size to 1 and trained it like that, everything worked fine. I'm wondering still if there is a way to train network with batch_size != 1 and then later freeze graph and reuse it by classifying images 1 by 1? Maybe somehow tell CNN to receive tensors of shape [None, height, width, channels] so when I freeze graph and run it from C++ it will work with only 1 example. My current solution could be to train Cifar10 model with batch_size = 128 and then in C++ create tensor with shape [128, height, width, channels], only first image would be real image and other "images" would be just random numbers, but that seems like waste of processing time and resources.

**Comment by girving** [comment]
@petewarden Can you take a look at this graph feezing issue?

**Comment by SeanTyco** [comment]
Was there any progress on this?

**Comment by girving** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
